File: plugins/devkit/scripts/workflow/pre-bash-gate.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");
const crypto = require("crypto");

function sanitizeSessionId(id) {
  if (!id) return crypto.randomUUID();
  const sanitized = id.replace(/[^a-zA-Z0-9-]/g, "");
  return sanitized || crypto.randomUUID();
}

function emitDecision(decision, reason, additionalContext) {
  const out = {
    hookSpecificOutput: {
      hookEventName: "PreToolUse",
      permissionDecision: decision,
      permissionDecisionReason: reason,
    },
  };
  if (additionalContext) {
    out.hookSpecificOutput.additionalContext = additionalContext;
  }
  process.stdout.write(JSON.stringify(out));
}

function findPluginJsonRelPath() {
  try {
    const { execSync } = require("child_process");
    const repoRoot = execSync("git rev-parse --show-toplevel", {
      encoding: "utf8",
      timeout: 5000,
      stdio: ["pipe", "pipe", "pipe"],
    }).trim();

    const candidates = [
      "plugins/devkit/.claude-plugin/plugin.json",
      ".claude-plugin/plugin.json",
    ];
    for (const c of candidates) {
      if (fs.existsSync(path.join(repoRoot, c))) return c;
    }
  } catch {
    // not in a git repo
  }
  return null;
}

function getGitVersion(ref, relPath) {
  try {
    const { execSync } = require("child_process");
    const output = execSync(`git show ${ref}:${relPath}`, {
      encoding: "utf8",
      timeout: 5000,
      stdio: ["pipe", "pipe", "pipe"],
    });
    const data = JSON.parse(output);
    return data.version || null;
  } catch {
    return null;
  }
}

function suggestVersion(currentVersion) {
  const { execSync } = require("child_process");
  const [major, minor, patch] = currentVersion.split(".").map(Number);

  try {
    const log = execSync("git log HEAD --format=%s -20", {
      encoding: "utf8",
      timeout: 5000,
    });
    const messages = log.trim().split("\n");

    let hasBreaking = false;
    let hasFeat = false;

    for (const msg of messages) {
      if (/^.*!:/.test(msg) || /BREAKING CHANGE/.test(msg)) hasBreaking = true;
      if (/^feat(\(.*\))?:/.test(msg)) hasFeat = true;
      if (/^chore.*bump version/.test(msg)) break;
    }

    if (hasBreaking)
      return { version: `${major + 1}.0.0`, reason: "BREAKING CHANGE 検出" };
    if (hasFeat)
      return {
        version: `${major}.${minor + 1}.0`,
        reason: "feat コミット検出",
      };
    return {
      version: `${major}.${minor}.${patch + 1}`,
      reason: "デフォルト patch",
    };
  } catch {
    return {
      version: `${major}.${minor}.${patch + 1}`,
      reason: "デフォルト patch",
    };
  }
}

function checkVersionBump() {
  const relPath = findPluginJsonRelPath();
  if (!relPath) {
    return { blocked: false };
  }

  const headVersion = getGitVersion("HEAD", relPath);
  const remoteVersion = getGitVersion("origin/main", relPath);

  if (!headVersion || !remoteVersion) {
    return { blocked: false };
  }

  if (headVersion === remoteVersion) {
    const recommended = suggestVersion(remoteVersion);
    return {
      blocked: true,
      message: `[devkit-workflow] ⛔ git push がブロックされました。plugin.json のバージョンが変更されていません（現在: ${remoteVersion}）。\n推奨バージョン: ${recommended.version}（理由: ${recommended.reason}）\nplugin.json の version を更新してからコミットし直してください。`,
    };
  }

  return { blocked: false };
}

function main() {
  let input = "";
  try {
    if (!process.stdin.isTTY) {
      input = fs.readFileSync(0, "utf8");
    }
  } catch {
    emitDecision("pass", "stdin not available");
    return;
  }

  if (!input) {
    emitDecision("pass", "no input");
    return;
  }

  let toolInput = {};
  let sessionId = "";
  try {
    const parsed = JSON.parse(input);
    toolInput = parsed.tool_input || {};
    sessionId = sanitizeSessionId(parsed.session_id || "");
  } catch {
    emitDecision("pass", "malformed input");
    return;
  }

  const command = toolInput.command || "";
  if (!command) {
    emitDecision("pass", "no command");
    return;
  }

  // Detect git commit or git push commands
  // Matches: git commit, git -C /path commit, git commit && ..., git commit; ..., etc.
  const isGitCommit = /\bgit\b.*\bcommit(?:\s|$|[;&|])/.test(command);
  const isGitPush = /\bgit\b.*\bpush(?:\s|$|[;&|])/.test(command);

  if (!isGitCommit && !isGitPush) {
    emitDecision("pass", "not a git commit/push");
    return;
  }

  // Check workflow state file for review completion
  const claudeDir = path.join(process.env.HOME || "", ".claude");
  const stateFile = path.join(
    claudeDir,
    `devkit-workflow-${sessionId}.json`
  );

  let state = null;
  try {
    if (fs.existsSync(stateFile)) {
      state = JSON.parse(fs.readFileSync(stateFile, "utf8"));
    }
  } catch {
    // state file unreadable
  }

  // If no state file exists, ask for confirmation
  if (!state) {
    const action = isGitCommit ? "commit" : "push";
    emitDecision(
      "ask",
      `git ${action} detected without workflow state`,
      `[devkit-workflow] \u26a0\ufe0f git ${action} \u304c\u691c\u51fa\u3055\u308c\u307e\u3057\u305f\u3002\u30ef\u30fc\u30af\u30d5\u30ed\u30fc\u72b6\u614b\u304c\u78ba\u8a8d\u3067\u304d\u307e\u305b\u3093\u3002\u30ec\u30d3\u30e5\u30fc\u30d5\u30a7\u30fc\u30ba\uff08Phase 5/7\uff09\u3092\u5b8c\u4e86\u3057\u3066\u3044\u307e\u3059\u304b\uff1f`
    );
    return;
  }

  const phasesPassed = Array.isArray(state.phases_passed)
    ? state.phases_passed
    : [];
  const hasReview =
    phasesPassed.includes("plan_review") &&
    phasesPassed.includes("impl_review");

  if (isGitCommit && !hasReview) {
    emitDecision(
      "ask",
      "git commit detected without review phase marker",
      `[devkit-workflow] \u26a0\ufe0f git commit \u304c\u691c\u51fa\u3055\u308c\u307e\u3057\u305f\u304c\u3001\u30ec\u30d3\u30e5\u30fc\u30d5\u30a7\u30fc\u30ba\u306e\u5b8c\u4e86\u30de\u30fc\u30ab\u30fc\u304c\u3042\u308a\u307e\u305b\u3093\u3002\n8\u30d5\u30a7\u30fc\u30ba\u30ef\u30fc\u30af\u30d5\u30ed\u30fc\u306e Phase 5\uff08\u8a08\u753b\u30ec\u30d3\u30e5\u30fc\uff09\u307e\u305f\u306f Phase 7\uff08\u5b9f\u88c5\u30ec\u30d3\u30e5\u30fc\uff09\u3092\u5b8c\u4e86\u3057\u3066\u304b\u3089\u30b3\u30df\u30c3\u30c8\u3057\u3066\u304f\u3060\u3055\u3044\u3002`
    );
    return;
  }

  if (isGitPush && !phasesPassed.includes("commit_review")) {
    emitDecision(
      "ask",
      "git push detected without commit review marker",
      `[devkit-workflow] \u26a0\ufe0f git push \u304c\u691c\u51fa\u3055\u308c\u307e\u3057\u305f\u304c\u3001\u30b3\u30df\u30c3\u30c8\u524d\u30ec\u30d3\u30e5\u30fc\uff08Phase 8 Step 2\uff09\u306e\u5b8c\u4e86\u30de\u30fc\u30ab\u30fc\u304c\u3042\u308a\u307e\u305b\u3093\u3002`
    );
    return;
  }

  // Version bump check for git push
  if (isGitPush) {
    const versionCheck = checkVersionBump();
    if (versionCheck.blocked) {
      emitDecision(
        "block",
        "plugin.json version not bumped",
        versionCheck.message
      );
      return;
    }
  }

  // All checks passed
  emitDecision("pass", "workflow checks passed");
}

main();

File: plugins/devkit/scripts/workflow/pre_bash_gate.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import re
import sys
from pathlib import Path


SCRIPT_DIR = Path(__file__).resolve().parent
sys.path.insert(0, str(SCRIPT_DIR))
sys.path.insert(0, str(SCRIPT_DIR.parent))

from check_plugin_version_bump import run_check
from workflow_state import normalize_phase_token, read_json, sanitize_session_id, state_file_for


def emit_decision(decision: str, reason: str, additional_context: str | None = None) -> None:
    payload: dict[str, object] = {
        "hookSpecificOutput": {
            "hookEventName": "PreToolUse",
            "permissionDecision": decision,
            "permissionDecisionReason": reason,
        }
    }
    if additional_context:
        payload["hookSpecificOutput"]["additionalContext"] = additional_context
    print(json.dumps(payload, ensure_ascii=False), end="")


def suggest_version_message() -> str | None:
    code, payload = run_check()
    if code == 0:
        return None
    if payload.get("reason") != "plugin version not bumped":
        return None
    return (
        "[devkit-workflow] ⛔ git push がブロックされました。"
        f"plugin.json のバージョンが変更されていません（現在: {payload.get('baseVersion')}）。\n"
        f"必要条件: {payload.get('required')}\n"
        "plugin.json の version を更新してからコミットし直してください。"
    )


def main() -> int:
    try:
        input_text = sys.stdin.read() if not sys.stdin.isatty() else ""
    except Exception:
        emit_decision("pass", "stdin not available")
        return 0

    if not input_text:
        emit_decision("pass", "no input")
        return 0

    try:
        parsed = json.loads(input_text)
    except json.JSONDecodeError:
        emit_decision("pass", "malformed input")
        return 0

    tool_input = parsed.get("tool_input") if isinstance(parsed, dict) else {}
    session_id = sanitize_session_id(str(parsed.get("session_id", ""))) if isinstance(parsed, dict) else ""
    command = tool_input.get("command", "") if isinstance(tool_input, dict) else ""
    if not command:
        emit_decision("pass", "no command")
        return 0

    is_git_commit = re.search(r"\bgit\b.*\bcommit(?:\s|$|[;&|])", command) is not None
    is_git_push = re.search(r"\bgit\b.*\bpush(?:\s|$|[;&|])", command) is not None
    if not is_git_commit and not is_git_push:
        emit_decision("pass", "not a git commit/push")
        return 0

    state = read_json(state_file_for(session_id))
    if not state:
        action = "commit" if is_git_commit else "push"
        emit_decision(
            "ask",
            f"git {action} detected without workflow state",
            f"[devkit-workflow] ⚠️ git {action} が検出されました。ワークフロー状態が確認できません。レビューゲート（Phase 5/7）を完了していますか？",
        )
        return 0

    phases_passed: list[str] = []
    raw_phases = state.get("phases_passed")
    if isinstance(raw_phases, list):
        for token in raw_phases:
            normalized = normalize_phase_token(token if isinstance(token, str) else None)
            if normalized:
                phases_passed.append(normalized)

    has_review = (
        "plan_review_completed" in phases_passed
        and "implementation_review_completed" in phases_passed
    )
    if is_git_commit and not has_review:
        emit_decision(
            "ask",
            "git commit detected without review phase marker",
            "[devkit-workflow] ⚠️ git commit が検出されましたが、レビューゲートの完了マーカーがありません。\nagent team ワークフローの Phase 5 または Phase 7 を完了してからコミットしてください。",
        )
        return 0

    if is_git_push and "commit_review_completed" not in phases_passed:
        emit_decision(
            "ask",
            "git push detected without commit review marker",
            "[devkit-workflow] ⚠️ git push が検出されましたが、コミット前確認（Phase 8 Step 2）の完了マーカーがありません。",
        )
        return 0

    if is_git_push:
        version_message = suggest_version_message()
        if version_message:
            emit_decision("block", "plugin.json version not bumped", version_message)
            return 0

    emit_decision("pass", "workflow checks passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
